import random
import subprocess


def read_from_file(filename):
    f = open(filename, 'r')
    content = f.read()
    f.close()
    return content


def write_on_file(filename, content):
    f = open(filename, 'w')
    f.write(content)
    f.close()


def calc_next_version(version):
    version = version.split('.')
    version[-1] = str(int(version[-1]) + 1)
    return '.'.join(version)


def calc_previous_version(version):
    version = version.split('.')
    version[-1] = str(int(version[-1]) - 1)
    return '.'.join(version)


def get_finished_fallback():
    return random.choice(['Prontinho, tudo finalizado.',
                          'Tudo certo, vocÃª jÃ¡ pode conferir a nova versÃ£o.',
                          'Estou super feliz, tudo deu certo!',
                          'Trabalho feito.',
                          'MissÃ£o comprida!',
                          'Pronto, tudo terminado.',])


def get_update_fallback(version):
    return random.choice([f'A nova versÃ£o {version} jÃ¡ estÃ¡ no ar! ðŸ¥³',
                          f'A versÃ£o {version} jÃ¡ estÃ¡ pronta e operando!',
                          f'A versÃ£o {version} estÃ¡ operando normalmente, como o esperado!',])


def get_error_fallback():
    return random.choice(['Vish, alguma coisa deu errado. ðŸ˜µ\nO erro Ã© esse aqui:',
                          'Algo deu errado... Segue o que aconteceu:',
                          'Eita, aconteceu alguma coisa que eu nÃ£o estava prevendo ðŸ¤”ðŸ˜³\nAqui estÃ¡ o erro: ðŸ‘‡ðŸ‘‡ðŸ‘‡ðŸ‘‡ðŸ‘‡ðŸ‘‡',
                          'Panico! Aconteceu um imprevisto...',
                          'Ehh, talvez vocÃª queira dar uma olhada nisso, deve ser importante: ðŸ‘‡',
                          'Algo de errado nÃ£o estÃ¡ certo\nDa uma olha nisso aqui:',
                          'Desculpa, mas parece que aconteceu alguma coisa:',])


def get_rollback_fallback(version):
    return random.choice([f'Eu voltei pra versÃ£o {version}.',
                          f'A versÃ£o {version} estÃ¡ executando novamente!',
                          f'E fica tranquilo a versÃ£o {version} jÃ¡ estÃ¡ executando!'])



def get_version_fallback(local, version):
    return random.choice([f'A versÃ£o atual do {local} Ã© a {version}.',
                          f'AtÃ© onde eu sei, a versÃ£o atual do {local} Ã© a {version}.',
                          f'Pelo que me lembro, a versÃ£o atual do {local} Ã© a {version}.'])


class Executor():
    def __init__(self, bot, user_id, verbose=False):
        self.bot = bot
        self.user_id = user_id
        self.verbose = verbose

    def log(self, message):
        #print("--------"*10, "\n", message)
        self.bot.send_message(self.user_id, f'> {message}')

    def message(self, message):
        self.bot.send_message(self.user_id, message)

    def error(self, message):
        # Something went wrong
        self.bot.send_message(self.user_id, get_error_fallback())
        self.bot.send_message(self.user_id, f'> {message}')

    def run_command(self, command):
        trace = ''

        # Exec command and get terminal output.
        proc = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        while proc.poll() is None:
            trace += proc.stdout.readline().decode('utf-8', 'ignore')

        # Log to bot.
        if self.verbose:
            self.log(trace)

        # Check for errors and raise then.
        if proc.wait() > 0:
            self.error(trace)
            raise Exception